### ISSUE TITLE
Lecture 4: Deep Neural Networks

### ISSUE THREAD

**Issue by KimDahye** [KimDahye/deeplearning_study#4]
## 4.1 Deep L-layer Neural network
- ref1: coursera w4 pdf (source: https://www.deeplearning.ai/)
- ref2: https://johngiorgi.github.io/deeplearning.ai-coursera-notes/neural_networks_and_deep_learning/week_4/

### Deep neural network: shallow vs deep
- hidden layer가 많을수록 deeper NN, hidden layer가 적을 수록 shallower NN.
  - Shallow NN is a NN with one or two layers.
  - Deep NN is a NN with three or more layers.
###  Notations
<img width="500" alt="Screen Shot 2021-09-08 at 8 13 48 PM" src="https://user-images.githubusercontent.com/6873655/132499679-b9a22ab6-4f5f-473a-badc-a766af369045.png">
<img width="500" alt="Screen Shot 2021-09-08 at 8 33 12 PM" src="https://user-images.githubusercontent.com/6873655/132502110-17016996-1fe7-4ddc-8114-b00a206346b0.png">

## 4.2 Forward Propagation in a Deep Network
<img width="500" alt="Screen Shot 2021-09-08 at 8 30 20 PM" src="https://user-images.githubusercontent.com/6873655/132501736-7e20c4ed-136b-4757-a3fa-6a3dfbdeee8a.png">
<img width="500" alt="Screen Shot 2021-09-08 at 8 34 49 PM" src="https://user-images.githubusercontent.com/6873655/132502323-7160198a-1c33-412e-81a2-ba96ca76cf65.png">

## 4.3 Getting your matrix dimensions right
- "충격, Andrew Ng의 딥러닝 디버깅 tool이 paper, pen 조합으로 밝혀져..." 
  - 그만큼 딥러닝 코딩시엔 matrix dimension이 헷갈릴 수 있다.
- 정돈된 수식
  <img width="500" alt="Screen Shot 2021-09-08 at 8 35 46 PM" src="https://user-images.githubusercontent.com/6873655/132502438-f42b1922-9140-4eb3-8891-beafd68245ae.png">
  - vectorized version에서 W, b의 dimension은 고정이고, X, Z, A만 (n, 1) -> (n, m)으로 변경되는 것 // column vector -> matrix

## 4.4 Why deep representations?
- [1] intuition: hidden layer가 많을수록 처음엔 기본적인 feature를 학습하고, 나중의 layer는 복잡한 feature를 학습하게 됨
  <img width="500" alt="Screen Shot 2021-09-08 at 8 38 07 PM" src="https://user-images.githubusercontent.com/6873655/132502749-d483a992-a390-4249-8ad6-485a03125175.png">
- [2] circuit theory:  deep neural network에서는 neuron 몇개만으로 계산할 수 있는데, hidden layer 수를 제한하면 neuron을 매우 많이 써서 계산해야하는 함수들이 존재함
 - deep NN일 때 더 학습이 잘될 수 있음 (조금만 - neuron 개수 - 학습하면 되니까)

## 4.5 Building blocks of deep neural netowrks
<img width="500" alt="Screen Shot 2021-09-08 at 8 42 44 PM" src="https://user-images.githubusercontent.com/6873655/132503292-2e1159e9-3abd-4e4e-bd4b-6da6eb294c0b.png">
<img width="500" alt="Screen Shot 2021-09-08 at 8 42 55 PM" src="https://user-images.githubusercontent.com/6873655/132503305-53c2d05f-540c-4b49-8e25-2f70474d8525.png">

## 4.6 Forward and backward propagation
<img width="500" alt="Screen Shot 2021-09-08 at 8 44 09 PM" src="https://user-images.githubusercontent.com/6873655/132503458-ef199e33-e07c-4fe3-86cd-962c4abdba6c.png">
<img width="500" alt="Screen Shot 2021-09-08 at 8 44 24 PM" src="https://user-images.githubusercontent.com/6873655/132503488-510f533f-07ab-40e6-9049-84c338ab4127.png">

## 4.7 Parameters vs Hyperparameters
- parameter: 딥러닝으로 학습해야 하는 값들 - W, b
- hyper parameter: 딥러닝이 파라미터를 학습할 때 영향을 주는 모델 셋팅값들 - learning rate, # iterations, # hidden layers L, # hidden units n^[1], n^[2], choice of activation function, and more!
  - 이 hyper parameter 값을 바꿔가면서 Idea -> code -> experiment  cycle 을 계속 도는 것! (It's very empirical process!) 

## 4.8 What does this have to do with the bran?
- 신경세포와 비슷하다는 analogy를 많이 썼으나, 그 analogy를 고정하면 딥러닝의 neuron과 신경세포는 다른 특징들을 많이 갖고 있기 때문에 그 차이점들이 발견될 때마다 어색해짐 -> 이제는 잘 안쓰임
- 수식 마지막으로 깔끔하게 정리!
  - <img width="500" alt="Screen Shot 2021-09-08 at 8 49 43 PM" src="https://user-images.githubusercontent.com/6873655/132504170-be49545d-ac11-4150-941b-f209c3835cd0.png">
